# Week 1 Hands-On Assignment

You are organizing your family contact list, but find that the current format of the phone numbers is flawed! The contact phone numbers are in different formats and you want valid phone numbers to all follow the same format.

Below, you can see that we have a file where each line consists of a person's name and their phone number.

First, run the code below to extract the lines.

In [1]:
text = None

with open('assets/phonebook.txt', 'r') as file:
    lines = file.readlines()

print(*lines[:5],sep='\n')  # print the first 5 rows in the file

David Smith: +1-361-789-5031

Michael Rodriguez: 801 998 5951

Sarah Johnson: (270) 573-8188

Jane Williams: 571.358.9802

Laura Davis: (965) 511-7358



# Task 1: Create reg expression

These are possible formats a phone number could be stored in:

Parentheses: (409) 339-9584

Dashes: 943-641-7270

Dots: 574.874.7163

Spaces: 175 335 3625

International: +1-536-234-1765

Plain: 6565865784

Create a 'reg' expression such that the given code is able to extract phone numbers in every one of these formats.

When we use reg.findall, we want to obtain a list of 3 strings containing the digits of each phone number separated by section. For example, if we have:

`str = John Smith: 123 456 7890`

We want `re.findall(reg, str)` on this line using our regex expression to obtain a list that contains the following:

['123','456','7890']

Look up how to use regex 'capturing groups' to accomplish this.

Assign your reg expression to `reg` in `task_1_solution`.

We use your reg expression in `extract_contents` to extract the contents of `lines` into a dictionary where each entry is formatted like so:

`'First Last': ['000', '000', '0000']`

In [2]:
import re


def extract_contents(reg):
    extracted_contents = {}
    for line in lines:
        line_content = line.split(':')
        matches = re.findall(reg, line_content[1])
        for match in matches:
            
            # Filter out empty strings and None values from the match
            parts = [part for part in match if part]
            extracted_contents[line_content[0]] = (parts)
    return extracted_contents

In [3]:

def task_1_solution():

    # Regular expression to match all the required formats
    reg_str = r"""
        (?:\+1[-\s\.]?)?                # optional country code
        (?:\(?(\d{3})\)?[-\s\.]?)       # area code
        (\d{3})                         # central office code
        [-\s\.]?                        # separator
        (\d{4})                         # line number
    """
    reg = re.compile(reg_str, re.VERBOSE)
    return extract_contents(reg), reg


In [4]:
# Use the following cell to explore your solution

task_1_solution()

({'David Smith': ['354', '838', '1778'],
  'Michael Rodriguez': ['441', '696', '9737'],
  'Sarah Johnson': ['358', '810', '4550'],
  'Jane Williams': ['820', '683', '4201'],
  'Laura Davis': ['499', '429', '4987'],
  'Jane Brown': ['957', '100', '5004'],
  'Chris Smith': ['268', '513', '5535'],
  'Sarah Garcia': ['750', '900', '6664'],
  'Michael Smith': ['698', '277', '2309'],
  'David Williams': ['533', '861', '6676'],
  'Sarah Jones': ['265', '588', '1120'],
  'Chris Garcia': ['485', '524', '1795'],
  'David Johnson': ['287', '112', '2937'],
  'Sarah Smith': ['451', '215', '3820'],
  'David Rodriguez': ['290', '406', '8054'],
  'Michael Williams': ['748', '144', '6517'],
  'Jane Johnson': ['365', '619', '3603'],
  'Alex Rodriguez': ['256', '532', '1513'],
  'John Miller': ['224', '361', '0505'],
  'Laura Brown': ['522', '156', '4663'],
  'David Garcia': ['377', '280', '6312'],
  'Sarah Wilson': ['301', '276', '9935'],
  'Jane Wilson': ['556', '850', '4384'],
  'Michael Johnson': ['6

In [5]:

print(f"AG tests")
stu_contents, stu_reg = task_1_solution()


print(f"Your answer:\n{stu_contents, stu_reg }")

assert isinstance(
    stu_contents, dict
), f"Your function should return a dictionary "

def test_format(testLine):
    test_content = testLine.split(':')
    number_components = stu_reg.findall(test_content[1])
    parts = None
    for component in number_components:
        # Filter out empty strings and None values from the match
        parts = [part for part in component if part]
    return parts

assert (
    test_format("Jane Doe: (123) 456-7890") == ['123', '456', '7890']
), f"Your reg expression does not read numbers in the format '(123) 456-7890' correctly "

assert (
    test_format("Jane Doe: 123-456-7890") == ['123', '456', '7890']
), f"Your reg expression does not read numbers in the format '123-456-7890' correctly "

assert (
    test_format("Jane Doe: 123.456.7890") == ['123', '456', '7890']
), f"Your reg expression does not read numbers in the format '123.456.7890' correctly "

assert (
    test_format("Jane Doe: 123 456 7890") == ['123', '456', '7890']
), f"Your reg expression does not read numbers in the format '123 456 7890' correctly "

assert (
    test_format("Jane Doe: 1-123-456-7890") == ['123', '456', '7890']
), f"Your reg expression does not read numbers in the format '1-123-456-7890' correctly "

assert (
    test_format("Jane Doe: 1234567890") == ['123', '456', '7890']
), f"Your reg expression does not read numbers in the format '1234567890' correctly "

# Hidden tests are located in this cell

AG tests
Your answer:
({'David Smith': ['354', '838', '1778'], 'Michael Rodriguez': ['441', '696', '9737'], 'Sarah Johnson': ['358', '810', '4550'], 'Jane Williams': ['820', '683', '4201'], 'Laura Davis': ['499', '429', '4987'], 'Jane Brown': ['957', '100', '5004'], 'Chris Smith': ['268', '513', '5535'], 'Sarah Garcia': ['750', '900', '6664'], 'Michael Smith': ['698', '277', '2309'], 'David Williams': ['533', '861', '6676'], 'Sarah Jones': ['265', '588', '1120'], 'Chris Garcia': ['485', '524', '1795'], 'David Johnson': ['287', '112', '2937'], 'Sarah Smith': ['451', '215', '3820'], 'David Rodriguez': ['290', '406', '8054'], 'Michael Williams': ['748', '144', '6517'], 'Jane Johnson': ['365', '619', '3603'], 'Alex Rodriguez': ['256', '532', '1513'], 'John Miller': ['224', '361', '0505'], 'Laura Brown': ['522', '156', '4663'], 'David Garcia': ['377', '280', '6312'], 'Sarah Wilson': ['301', '276', '9935'], 'Jane Wilson': ['556', '850', '4384'], 'Michael Johnson': ['676', '905', '1229'], 'Al

# Task 2: Normalizing Data

Now, we'd like to normalize the phone numbers to the format: (123) 456-7890.

The function below will take in the ```stu_contents``` that you helped define in task 1 as an argument. It will then alter every entry in the ```stu_contents``` dictionary to be a string in the (123) 456-7890 format. The function will return the modified ```extracted_contents``` dictionary.

For example, 

The first entry in your task 1 `extracted_contents` is 

'David Smith': ['838', '177', '8461']

This should be turned into

'David Smith': '(838) 177-8461'

In [6]:
import copy


In [7]:

def task_2_solution():

    extracted_contents, _ = task_1_solution()

    extracted_contents = copy.deepcopy(extracted_contents)
    
    # YOUR CODE HERE
    for name, parts in extracted_contents.items():
        if len(parts) == 3:
            area, mid, last = parts
            formatted = f"({area}) {mid}-{last}"
            extracted_contents[name] = formatted
        else:
            # If the phone number format was invalid and didn't capture 3 parts
            extracted_contents[name] = "Invalid Number"

    return extracted_contents

In [8]:
# Use the following cell to explore your solution

task_2_solution()

{'David Smith': '(354) 838-1778',
 'Michael Rodriguez': '(441) 696-9737',
 'Sarah Johnson': '(358) 810-4550',
 'Jane Williams': '(820) 683-4201',
 'Laura Davis': '(499) 429-4987',
 'Jane Brown': '(957) 100-5004',
 'Chris Smith': '(268) 513-5535',
 'Sarah Garcia': '(750) 900-6664',
 'Michael Smith': '(698) 277-2309',
 'David Williams': '(533) 861-6676',
 'Sarah Jones': '(265) 588-1120',
 'Chris Garcia': '(485) 524-1795',
 'David Johnson': '(287) 112-2937',
 'Sarah Smith': '(451) 215-3820',
 'David Rodriguez': '(290) 406-8054',
 'Michael Williams': '(748) 144-6517',
 'Jane Johnson': '(365) 619-3603',
 'Alex Rodriguez': '(256) 532-1513',
 'John Miller': '(224) 361-0505',
 'Laura Brown': '(522) 156-4663',
 'David Garcia': '(377) 280-6312',
 'Sarah Wilson': '(301) 276-9935',
 'Jane Wilson': '(556) 850-4384',
 'Michael Johnson': '(676) 905-1229',
 'Alex Miller': '(879) 900-6029',
 'Emily Johnson': '(250) 964-3944',
 'Katie Garcia': '(185) 162-4248',
 'Katie Miller': '(443) 919-6759',
 'Chris

In [9]:

print(f"AG tests")
stu_ans = task_2_solution()


print(f"Your answer:\n{stu_ans}")

assert isinstance(
    stu_ans, dict
), f"Your function should return a dictionary "

# Hidden tests are located in this cell

del stu_ans


AG tests
Your answer:
{'David Smith': '(354) 838-1778', 'Michael Rodriguez': '(441) 696-9737', 'Sarah Johnson': '(358) 810-4550', 'Jane Williams': '(820) 683-4201', 'Laura Davis': '(499) 429-4987', 'Jane Brown': '(957) 100-5004', 'Chris Smith': '(268) 513-5535', 'Sarah Garcia': '(750) 900-6664', 'Michael Smith': '(698) 277-2309', 'David Williams': '(533) 861-6676', 'Sarah Jones': '(265) 588-1120', 'Chris Garcia': '(485) 524-1795', 'David Johnson': '(287) 112-2937', 'Sarah Smith': '(451) 215-3820', 'David Rodriguez': '(290) 406-8054', 'Michael Williams': '(748) 144-6517', 'Jane Johnson': '(365) 619-3603', 'Alex Rodriguez': '(256) 532-1513', 'John Miller': '(224) 361-0505', 'Laura Brown': '(522) 156-4663', 'David Garcia': '(377) 280-6312', 'Sarah Wilson': '(301) 276-9935', 'Jane Wilson': '(556) 850-4384', 'Michael Johnson': '(676) 905-1229', 'Alex Miller': '(879) 900-6029', 'Emily Johnson': '(250) 964-3944', 'Katie Garcia': '(185) 162-4248', 'Katie Miller': '(443) 919-6759', 'Chris Jones